src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersReservedSlotsBehavior.cs

```diff
@@ -0,0 +1,150 @@
+
+      /// <summary>
+      /// The proven header peers reserved slots threshold (%).
+      /// Represents the percentage of maximum connectable peers that we reserve for peers that are able to serve proven headers.
+      /// </summary>
+      private const decimal ProvenHeaderPeersReservedSlotsThreshold = 0.4M;
+
+      /// <summary>Instance logger.</summary>
+      private readonly ILogger logger;
+
+      public ProvenHeadersReservedSlotsBehavior(
+         IConnectionManager connectionManager,
+         ILoggerFactory loggerFactory,
+         ConsensusSettings consensusSettings,
+         ICheckpoints checkpoints,
+         IChainState chainState)
+      {
+         this.connectionManager = connectionManager;
+         this.loggerFactory = loggerFactory;
+         this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+         this.consensusSettings = consensusSettings;
+         this.checkpoints = checkpoints;
+         this.chainState = chainState;
+      }
+
+      /// <summary>
+      /// Processes and incoming message from the peer.
+      /// </summary>
+      /// <param name="peer">Peer from which the message was received.</param>
+      /// <param name="message">Received message to process.</param>
+      protected async Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+      {
+         switch (message.Message.Payload)
+         {
+            case VersionPayload version:
+               await this.ProcessVersionAsync(peer, version).ConfigureAwait(false);
+               break;
+         }
+      }
+
+      /// <summary>
+      /// Processes "version" message received from the peer.
+      /// Ensures we leave some available connection slots to nodes that are enabled to serve Proven Headers, if the number of minimum PH enabled nodes connected isn't met.
+      /// </summary>
+      /// <param name="peer">Peer from which the message was received.</param>
+      /// <param name="version">Payload of "version" message to process.</param>
+      private Task ProcessVersionAsync(INetworkPeer peer, VersionPayload version)
+      {
+         bool isBelowLastCheckpoint = this.consensusSettings.UseCheckpoints && (this.chainState.ConsensusTip.Height <= this.checkpoints.GetLastCheckpointHeight());
+
+         // We enforce having free slots for PH enabled nodes only when we have passed last checkpoint.
+         if (!isBelowLastCheckpoint)
+         {
+            PeerConnectorDiscovery connector = this.connectionManager.PeerConnectors.OfType<PeerConnectorDiscovery>().FirstOrDefault();
+            // If PeerConnectorDiscovery is not found means we are using other ways to connect peers (like -connect) and thus we don't enforce the rule.
+            if (connector != null)
+            {
+               int slotsReservedForProvenHeaderEnabledPeers = (int)Math.Round(connector.MaxOutboundConnections * ProvenHeaderPeersReservedSlotsThreshold, MidpointRounding.ToEven);
```

Comment:
please use shorter name and add comment to provide info you currently have in variable name

Reply:
funny enough i pointed this out in Slack #fullnode channel before using it
https://stratisgroupworkspace.slack.com/archives/C9TP7RP2R/p1539269982000100

>does `decimal slotsReservedForProvenHeaderEnabledPeers` sounds too verbose? :nerd_face:

and i just got 2 ACK about this.

Reply:
>please use shorter name  
 
What why is this a problem?